src/keyutil/keyutil.c

```diff
@@ -277,11 +291,115 @@ static int verb_public(int argc, char *argv[], void *userdata) {
+
+        if (!arg_output)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "--output= must be specified");
+
+        if (argc < 2)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Input and output must be specified");
+
+        if (arg_certificate_source_type == OPENSSL_CERTIFICATE_SOURCE_FILE) {
+                r = parse_path_argument(arg_certificate, /*suppress_root=*/ false, &arg_certificate);
+                if (r < 0)
+                        return r;
+        }
+
+        r = openssl_load_x509_certificate(
+                        arg_certificate_source_type,
+                        arg_certificate_source,
+                        arg_certificate,
+                        &certificate);
+        if (r < 0)
+                return log_error_errno(r, "Failed to load X.509 certificate from %s: %m", arg_certificate);
+
+        r = read_full_file(argv[1], &pkcs1, &pkcs1_len);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read PKCS1 file %s: %m", argv[1]);
+        if (pkcs1_len == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "PKCS1 file %s is empty", argv[1]);
+
+        /* Create PKCS7_SIGNER_INFO using X509 pubkey/digest NIDs */
+
+        _cleanup_(PKCS7_SIGNER_INFO_freep) PKCS7_SIGNER_INFO *signer_info = PKCS7_SIGNER_INFO_new();
+        if (!signer_info)
+                return log_oom();
+
+        if (!ASN1_INTEGER_set(signer_info->version, 1))
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set ASN1 integer: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+        if (!X509_NAME_set(&signer_info->issuer_and_serial->issuer, X509_get_issuer_name(certificate)))
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set issuer name: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+        ASN1_INTEGER_free(signer_info->issuer_and_serial->serial);
+        if (!(signer_info->issuer_and_serial->serial = ASN1_INTEGER_dup(X509_get0_serialNumber(certificate))))
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set issuer serial: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        int x509_mdnid = 0, x509_pknid = 0;
+        if (!X509_get_signature_info(certificate, &x509_mdnid, &x509_pknid, /* secbits= */ NULL, /* flags= */ NULL))
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to get X509 digest NID/PK: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        X509_ALGOR_set0(signer_info->digest_alg, OBJ_nid2obj(x509_mdnid), V_ASN1_NULL, /* pval= */ NULL);
+        X509_ALGOR_set0(signer_info->digest_enc_alg, OBJ_nid2obj(x509_pknid), V_ASN1_NULL, /* pval= */ NULL);
+
+        /* Create new PKCS7 using X509 certificate */
+
+        _cleanup_(PKCS7_freep) PKCS7 *pkcs7 = PKCS7_new();
```

Comment:
Ditto, PKCS7 variables are usually called `p7`

Reply:
as above

Reply:
@bluca Every openssl example and its internal code calls this `p7` though. Feels a bit weird to deviate from that, but OK